### ISSUE TITLE
BUG: interpolate.interpn fails with read only input

### ISSUE THREAD

**Issue by braingram** [scipy/scipy#17716]
### Describe your issue.

Using a slightly modified version of the example from the documentation
```python
import numpy
from scipy.interpolate import interpn


def value_func_3d(x, y, z):
    return 2 * x + 3 * y - z


x = numpy.linspace(0, 4, 5)
y = numpy.linspace(0, 5, 6)
z = numpy.linspace(0, 6, 7)
points = (x, y, z)
values = value_func_3d(*numpy.meshgrid(*points, indexing='ij'))
point = numpy.array([2.21, 3.12, 1.15])
print(interpn(points, values, point))
for ax in points:
    ax.flags.writeable = False
print(interpn(points, values, point))
```

Fails with
```bash
...
    return find_indices(self.grid, xi)
  File "_rgi_cython.pyx", line 95, in scipy.interpolate._rgi_cython.find_indices
  File "stringsource", line 660, in View.MemoryView.memoryview_cwrapper
  File "stringsource", line 350, in View.MemoryView.memoryview.__cinit__
ValueError: buffer source array is read-only
```

A more minimal example is:
```python
import numpy
from scipy.interpolate import _rgi


grid = (numpy.array([0, 1], dtype='f8'), )
xi = numpy.array([[0.5, 0.5], ])
grid[0].flags.writeable = False
print(_rgi.find_indices(grid, xi))
```

This may be related to https://github.com/scipy/scipy/issues/17172 and may have been introduced with the cython code in #17291

Adding a const to this line appear to fix the problem locally: https://github.com/scipy/scipy/blob/7b858b70c16c5559c2155a6bf2c39b32e45fbb78/scipy/interpolate/_rgi_cython.pyx#L79

### Reproducing Code Example

```python
import numpy
from scipy.interpolate import _rgi


grid = (numpy.array([0, 1], dtype='f8'), )
xi = numpy.array([[0.5, 0.5], ])
grid[0].flags.writeable = False
print(_rgi.find_indices(grid, xi))
```

### Error message

```shell
File "_rgi_cython.pyx", line 95, in scipy.interpolate._rgi_cython.find_indices
  File "stringsource", line 660, in View.MemoryView.memoryview_cwrapper
  File "stringsource", line 350, in View.MemoryView.memoryview.__cinit__
ValueError: buffer source array is read-only
```

### SciPy/NumPy/Python version information

1.11.0.dev0+0.057cf5c 1.24.1 sys.version_info(major=3, minor=10, micro=6, releaselevel='final', serial=0)

**Comment by braingram** [comment]
I don't have minimal cases yet for these but there may be related issues with non-c-contiguous arrays.
https://github.com/asdf-format/asdf/actions/runs/3833531489/jobs/6525091201#step:9:1460
and endianess (although this one is raised in `evaluate_linear_2d`):
https://github.com/asdf-format/asdf/actions/runs/3833531489/jobs/6525091201#step:9:1609

**Comment by ev-br** [comment]
Oh well. I'd be great if you could distill it to what can be used as unit tests. 
As far as fixes go, I wonder how large are the data you're working with. Can you afford a call to `np.ascontiguousarray` to copy data so that cython code can keep being `::1`? 

Also going forward I wonder how we (a collective "we" here) could catch these sort of things before a SciPy release. 

Any chance you could set up a test with nightly wheels?

**Comment by braingram** [comment]
Thanks!

Here's a minimal case for the non-c-contiguous issue:
```python
import numpy
from scipy.interpolate import interpn


x = numpy.linspace(0, 4, 5)
x = numpy.vstack((x, numpy.empty_like(x))).T.copy()[:, 0]
y = numpy.linspace(0, 5, 6)
z = numpy.linspace(0, 6, 7)
values = numpy.ones((5, 6, 7))
point = numpy.array([2.21, 3.12, 1.15])

points = (x, y, z)
print(interpn(points, values, point))
```
raises an error on 1.10 (not on 1.9)
```bash
    return find_indices(self.grid, xi)
  File "_rgi_cython.pyx", line 95, in scipy.interpolate._rgi_cython.find_indices
  File "stringsource", line 660, in View.MemoryView.memoryview_cwrapper
  File "stringsource", line 350, in View.MemoryView.memoryview.__cinit__
ValueError: ndarray is not C-contiguous
```

I'm not sure what's the best fix here. As you mentioned `np.ascontiguousarray` could be costly and I'm not that familiar with all the uses in the jwst code to know where this might be problematic.

Should I open a new issue for the non-c-contiguous error (and perhaps another when I can sort out the endianess error) or roll them all into this one?

Testing against nightly builds is a great suggestion. I don't think that is currently being done for the jwst package but @zacharyburnett and @williamjamieson know more than I do.

**Comment by ev-br** [comment]
From the scipy perspective, an ideal fix would be that 
- the fast path does ascountguousarray
- those users who are memory constrained would use the 'duck-typed' code path, https://github.com/scipy/scipy/blob/main/scipy/interpolate/_rgi.py#L253

The latter was originally added quoting out-of-memory datasets, hdf5  or whatnot; would be great to identify at least one user of this (or if there aren't any, we'll happily rip out all this complexity).

**Comment by ev-br** [comment]
Re same PR or separate: either way works. If you lump all fixes together, great. 
There doesn't seem to be a pressing need for a quick 1.10.1 release (just yet :-)), so there is no big rush.

**Comment by WilliamJamieson** [comment]
Testing against the nightly `scipy` builds was something I thought JWST and GWCS was doing. However, this was an oversight. I believe @zacharyburnett was planning on adding `scipy` (and `numpy`) nightly builds to the `dev` version testing for JWST.

**Comment by ev-br** [comment]
Actually, now that I look at this code again, the problem is only `grid_i` which is not a big deal memory-wise. So I guess the fix is to ensure contiguity in this line: https://github.com/scipy/scipy/blob/main/scipy/interpolate/_rgi.py#L20

The rest of the Cython code does not require contiguity, all other memoryviews are `[:, :]`. 

Going forward, it would be great to actually see the memory requirement. There is one other enhancement in the pipeline where C contiguity of `values` would be a big help. While this won't arrive tomorrow, it'd be good to both understand the constraints before rolling it in, and have a way to iron out the downstream usage wrinkles before it's released so we don't scramble the next day after 1.11 :-).

**Comment by braingram** [comment]
I've been looking into this a bit more and as you pointed out, fixing the c-contiguity prior to `grid.append` here appears to fix one of the issues:
https://github.com/scipy/scipy/blob/7b858b70c16c5559c2155a6bf2c39b32e45fbb78/scipy/interpolate/_rgi.py#L21-L31

However, it revealed another issue with read-only memory views this time in `evaluate_linear_2d` https://github.com/scipy/scipy/blob/7b858b70c16c5559c2155a6bf2c39b32e45fbb78/scipy/interpolate/_rgi_cython.pyx#L19-L23
The `values` either need to be writeable or the argument needs a `const`. However, adding `const` fails I think because it is a fused type. I'm using cython 0.29.32 and support for this might not be included until 0.29.33 https://github.com/cython/cython/pull/5076

I'm inclined to add a check for `self.values.flags.writeable` at the branch for the fast path until cython supports const fused type unless I'm missing something (it's been years since I touched cython so I wouldn't be surprised).
https://github.com/scipy/scipy/blob/7b858b70c16c5559c2155a6bf2c39b32e45fbb78/scipy/interpolate/_rgi.py#L332-L333

**Comment by ev-br** [comment]
Oh wow, cython _is_ going to fix the readonly fused memoryviews thing. That's great news indeed! 
Branching on `values.flags.writeable` sounds good to me, please add a comment so that we can find it and undo it when `const double_or_complex[:, ::1]`  works in cython.
